### ISSUE TITLE
The extension slows page loading in Firefox

### ISSUE THREAD

**Issue by fturcheti** [lusakasa/saka-key#30]
When the extension is enabled, general page loading times are greater. This is mostly noticeable because of the deferred loading of styles, as seen in the screen recording below.

Screen recording: http://d.pr/v/mJACd

In this screen recording, I first reload the page 2x with Saka Keys enabled and then 2x with the extension disabled. With Saka Keys enabled we can see the HTML structure without the styles applied and after a flicker the proper page layout. With Saka Keys disabled we get the proper page layout with the styles applied without any flickering.

For reference, I'm using Firefox Nightly 57.0a1 and Saka Key 1.20.0rc.

**Comment by eejdoowad** [comment]
Do you know of any way to mitigate this issue? Saka Key must run as soon as possible to install  keyboard event listeners before web pages can. I presume rendering is delayed because the page is busy executing Saka Key's code instead. The flickering appears to be Firefox only.

An idea I had is to use a small "loader script" to install event listeners early on, save any events to a queue, delay loading the full client until the page is rendered, then empty the event queue into the full client. This most natural way to send events from the "loader script" to the full client is by saving them to the window object. I actually use a "loader script/full client"  for the Chrome build, but v1.20.0 removed this from the Firefox build because Firefox inexplicably reset the window object on some pages, see #26.

**Comment by fturcheti** [comment]
Sorry, I don't know how to help with possible solutions. 

I tested versions 1.19.0rc and 1.18.3rc from the dev channel and both present the same issue, so it seems to be unrelated to the changes commited to version 1.20.0rc.

**Comment by eejdoowad** [comment]
The architecture existed in those builds, but I didn't delay loading the full client at all. I loaded it as soon as possible.

**Comment by fturcheti** [comment]
I don't see this issue when using Vimium-FF, so maybe you could take a look on its strategy to load scripts.

**Comment by eejdoowad** [comment]
Good point. Vimium actually loads all its scripts into every frame, so if my conjecture about this issue were true then Vimium should have exhibited it as well. Which leads me to believe, it might be something else. I'll try removing code that modifies the DOM on startup and see if there's any improvement.

**Comment by fturcheti** [comment]
Good, @eejdoowad. If you'd like me to test something for you, let me know.

**Comment by fturcheti** [comment]
After more tests I have seen this occur even with the addon disabled, so that's unclear what may be causing this issue, but it's certainly not originated by this extension.

**Comment by Stebalien** [comment]
I'm experiencing this issue as well and it does go away when I disable saka-key. It ocationally delays website CSS (especially on GitHub) but almost always delays custom stylus userstyles.

**Comment by eejdoowad** [comment]
I've since discovered that I'm performing cleanup in firefox by listening for the 'unload' event, which prevents caching. There's another event in firefox someone told me about (I can't quite recall which one) that also gets fired when a page unloads that I can use instead.

I need to get back on this.

**Comment by Stebalien** [comment]
That may be the `onbeforeunload` event. However, I'm assuming all the cleanup you have to do relates to the background page (as the actual page is going to be unloaded. If so, can't you use the tabs API? IMO, the less done in each content script the better (the less JavaScript has to be executed on each page load).

**Comment by fturcheti** [comment]
I'm using Vimium-FF for a while and I haven't seen this frequent delay happening with it. I tried the last Saka Key build and confirmed that this delay is still present. 

As the extended tests proved this is something related to Saka, I'm reopening the issue.

**Comment by eejdoowad** [comment]
The event is actually [pagehide](https://developer.mozilla.org/en-US/Firefox/Releases/1.5/Using_Firefox_1.5_caching#pagehide_event).

State about each open tab is maintained in the background page. This state must be cleaned up whenever the user navigates to a new page. In Chrome, this is done when the background page receives the port.onDisconnect event. In Firefox, this event is only triggered when a tab is closed. Thus, in Firefox, each page must notify the background page when it navigates. Page's no when they're being unloaded by listening for the unload event. Unfortunately, listening for this event disables caching. Fortunately, firefox supports the pagehide event, which I'll switch to.

A workaround wouldn't be needed if this [bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1223425#c4) were resolved.

**Comment by MightyPork** [comment]
Firefox 60, was investigating FOUC by disabling individual extensions and found that saka-keys is the culprit. So this issue is definitely still relevant

**Comment by umpirsky** [comment]
In https://github.com/lusakasa/saka-key/issues/208 I can reproduce and the issue is gone after disabling saka key addon. So yes, still relevant.

**Comment by neilmacintyre** [comment]
@eejdoowad Can your give example of when this is true?

> Saka Key must run as soon as possible to install keyboard event listeners before web pages can. 

When I simply change the run_at parameter to in the Firefox manifest JSON to document_idle, the issue is resolved; of course this goes against what you said above. At the moment I can't find any websites where this causes an issue; specific examples would be much appreciated.

**Comment by Arthaey** [comment]
After I installed Saka Key, I started noticing that _all_ websites started loading noticeably slower, with a distinct "flash" (more like a full second or two!) of unstyled content.

@neilmacintyre How did you edit the manifest.json file? Is there a way to do it without having to load a "Temporary Extension" via the Debug Add-ons Firefox page every time I restart my browser?

versions: Saka-Key 1.25.5, Firefox 72.0b11 (developer edition)

**Comment by alexherbo2** [comment]
Yes.

[Firefox – Extension Signing – FAQ]

[Firefox – Extension Signing – FAQ]: https://wiki.mozilla.org/Add-ons/Extension_Signing#FAQ
